docs/flip/1777-default-dtype.md

```diff
@@ -0,0 +1,102 @@
+# FLIP: Default dtypes
+
+
+- Start Date: 2022-01-11
+- FLIP PR: [#1776](https://github.com/google/flax/pull/1776)
+- FLIP Issue: [#1777](https://github.com/google/flax/issues/1777)
+
+
+## Summary
+
+This FLIP proposes to replace the default dtype which is currently hard-coded to float32, and instead use the JAX type promotion results to derive a default dtype from the input and parameters of a layer.
+
+
+## Motivation
+
+Currently, linen Modules always produce float32 outputs regardless of input and parameter dtypes. Half-precision types like float16 and bfloat16 are supported by explicitly passing the half-precision type to each layer. The way this is currently implemented is that each layer has a dtype argument with float32 as the default value. The layer guarantees that this dtype will be the return type of the result returned by `__call__`.
+
+The current behavior is problematic and results in silent bugs especially for dtypes that do not fit inside float32 (complex, float64). Also the linen dtype behavior is significantly different from how NumPy and by extension JAX handle dtypes.
+
+
+### Dtypes in JAX
+
+JAX uses a NumPy inspired [dtype promotion](https://github.com/google/jax/blob/main/jax/_src/dtypes.py) mechanism as explained [here](https://jax.readthedocs.io/en/latest/type_promotion.html?highlight=lattice#type-promotion-semantics). The type promotion rules are summarized by the following type lattice:
+
+![JAX type promotion lattice](https://jax.readthedocs.io/en/latest/_images/type_lattice.svg)
+
+
+
+## Dtypes in Linen
+
+Beside input arguments, state and in particular parameters could affect dtype promotion. For example: We might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.
```

Comment:
```suggestion
Beside input arguments, state and in particular parameters could affect dtype promotion. For example: we might feed a float64 input to a Dense layer with float32 parameters. Currently the result would be truncated to float32. If the input is a complex number the result is even worse because the imaginary part will be silently dropped when casting to float32.
```